woocommerce/google-listings-and-ads
# Pre-push hook to run PHPCS

### Changes proposed in this Pull Request:

I can't be the only one [always](https://github.com/woocommerce/google-listings-and-ads/pull/386/commits/51d1f299968be0907736761b9af02cd2f72ef8c9) [pushing](https://github.com/woocommerce/google-listings-and-ads/commit/7a8ece9b22fa38537d782463f992fe80ad12c7f1) [PHPCS](https://github.com/woocommerce/google-listings-and-ads/commit/d552add003ed356dd44173bb62e79fb64a0ac450) [commits](https://github.com/woocommerce/google-listings-and-ads/commit/da919f9ff79241eae82a170f2607a97eb3c6083b) after the fact, right?

This adds a pre-push hook to run PHPCS (on the whole branch for newly-pushed branches, and only on the changed files for existing branches). Borrowed from [this gist](https://gist.github.com/pfrenssen/498fc52fea3f965f6640).

The only issue is that it runs PHPCS against the files in the working directory, so if there's an issue in the committed files and you fix it without committing the fix, the push will go through. It's possible to enable [stashing changes](https://github.com/woocommerce/google-listings-and-ads/pull/387/commits/3b2d94e3b45861e35aef4ba53e42b5e10c3b6a43?diff=unified&w=1) before running PHPCS, but that unstages any changes and that may not be ideal either 🤷🏻 

### Detailed test instructions:

1. Run `composer i`
    - Confirm that the pre-push hook is added at the end of the process:
    `Linking pre-push hook script`
2. Create a new branch off of this one `git checkout -b pr-test/testing`
2. Edit a file and "break" it with regards to PHPCS
    - Try mis-aligning the parameter names in the PHPDoc for `Installer::__construct()`
3. Commit the file `git commit -a -m "Test bad phpcs"`
4. Try to push the new branch `git push --set-upstream origin pr-test/testing`
    - Confirm that PHPCS is run against ALL files and fails, preventing push
5. Fix the files (`vendor/bin/phpcbf src/Installer.php`) and commit `git commit -a -m "Test fixed phpcs"`
6. Then push again
    - Confirm the push works if the PHPCS sniffer finds no errors
6. Make the same "breaking" changes again and commit them `git commit -a -m "Test bad phpcs"`
7. Commit the new changes `git push`
   - Confirm that PHPCS runs only on the changed file and fails, preventing push
8. Fix the broken file and commit the fix.
   - `vendor/bin/phpcbf src/Installer.php` and `git commit -a -m "Test fixed phpcs"`
9. Push again `git push`
    - Confirm the push works if the PHPCS sniffer finds no errors
9. Disable the hook by deleting `.git/hooks/pre-push` (to "force" push broken files if necessary)=
1. Confirm that you can commit and push files with PHPCS errors.
    - Make the same breaking changes, commit and push.
1. Run `composer -i` again to reinstall the pre-push hook
1. ❗ Remember to delete the test branch from the repo: <https://github.com/woocommerce/google-listings-and-ads/branches/yours>
